### ISSUE TITLE
1.37: Using nvm collides with npm script executing within code

### ISSUE THREAD

**Issue by shiftgeist** [microsoft/vscode#78784]
- VSCode Version: 1.37
- OS Version: Debian 10

Steps to Reproduce:

1. Update code to 1.37
2. Open "NPM SCRIPTS" in explorer sidebar menu
3. First it didn't found node:
```
/usr/bin/env: „node“: Datei oder Verzeichnis nicht gefunden
```

4. My fix attempt:
```
ln -s $(echo $NVM_BIN) /usr/bin/node
```

5. But than it had no permission:
```
/usr/bin/env: „node“: Keine Berechtigung
```

6. Fin. Here is where I have no clue how to continue.

Does this issue occur when all extensions are disabled?: Yes

**Comment by egamma** [comment]
Need to understand your setup better. 

Do you have 'node' on your PATH and you can run 'npm run SOMESCRIPT' from the command line?

**Comment by shiftgeist** [comment]
Yes `npm run` works as expected.

**Comment by egamma** [comment]
hmm I have the same nvm setup

![image](https://user-images.githubusercontent.com/172399/63784388-def6d200-c8ee-11e9-9ff4-0729b1362f42.png)

I suspect it has to do with the definition of your Shells startup scripts. What shell are you using and where are you adding nvm to the path in there. One thing to keep in mind is that tasks do not launch an interactive shell.

**Comment by shiftgeist** [comment]
I'm using zsh (`/usr/bin/zsh`). I'm not shure what you mean by `where are you adding nvm to the path in there`. In my .zshrc there is the default nvm loader included:

```
export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh" # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
```

Location of my node version (handled by nvm, i think :thinking:):

![image](https://user-images.githubusercontent.com/30648730/63834346-6dae3200-c975-11e9-8f09-17de823450c5.png)

I tried adding only the /bin path to $PATH, like you showed in [#27185](https://github.com/microsoft/vscode/issues/27185#issue-230956687) (without `/node`)
```
PATH="$NVM_DIR/versions/node/v12.9.1/bin:$PATH"
```
That wasn't successful.

```
> Executing task: npm run start <

zsh:1: command not found: npm
The terminal process terminated with exit code: 127

Terminal will be reused by tasks, press any key to close it.
```

**Comment by egamma** [comment]
I'm also using zsh.

>How did you executing a ShellTask from the API?

You have to implement a small extension to do so, but the API has changed.

@alexr00 any suggestion how to track this down?

**Comment by alexr00** [comment]
Tasks don't run as an interactive shell or as a login shell. But the VS Code terminal does try to pass through the environment that VS Code has. How are you launching VS Code?

One thing that might help is seeing what environment your tasks are actually getting. Can you try the following:

script.js in a folder open in VS Code.
```
console.log(process.env);
```

Task. If your task can't find `node` you'll need to use the full path for it.
```
        {
            "type": "shell",
            "command": "node script.js",
            "problemMatcher": [],
            "label": "Print env"
        }
```

Run the task and share the output. Feel free remove any sensitive environment entries. I just want to make sure that your environement looks close enough.

**Comment by shiftgeist** [comment]
# Task definition

Alright. So I had to make some adjustments to run the script in vscode:

```diff
{
    "version": "2.0.0",
    "tasks": [
        {
            "type": "shell",
-           "command": "node script.js",
+           "command": "/home/fhungenberg/.nvm/versions/node/v12.9.1/bin/node script.js",
            "problemMatcher": [],
            "label": "Print env"
        }
    ]
}
```
Because else:

```bash
> Executing task: node script.js <

zsh:1: command not found: node
The terminal process terminated with exit code: 127

Terminal will be reused by tasks, press any key to close it.
```

# Outputs (sorted alphabetically)

Running the task in vscode:

```
{
  _: '/home/fhungenberg/.nvm/versions/node/v12.9.1/bin/node',
  APPLICATION_INSIGHTS_NO_DIAGNOSTIC_CHANNEL: 'true',
  BAMF_DESKTOP_FILE_HINT: '/var/lib/snapd/desktop/applications/code_code.desktop',
  COLORTERM: 'truecolor',
  DBUS_SESSION_BUS_ADDRESS: 'unix:path=/run/user/1000/bus',
  DESKTOP_SESSION: 'gnome-classic',
  DISABLE_WAYLAND: '1',
  DISPLAY: ':0',
  GDK_PIXBUF_MODULE_FILE: '/home/fhungenberg/snap/code/common/.cache/gdk-pixbuf-loaders.cache',
  GDK_PIXBUF_MODULEDIR: '/snap/code/13/usr/lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders',
  GDM_LANG: 'de_DE.UTF-8',
  GDMSESSION: 'gnome-classic',
  GIO_LAUNCHED_DESKTOP_FILE_PID: '14683',
  GIO_LAUNCHED_DESKTOP_FILE: '/var/lib/snapd/desktop/applications/code_code.desktop',
  GJS_DEBUG_OUTPUT: 'stderr',
  GJS_DEBUG_TOPICS: 'JS ERROR;JS LOG',
  GNOME_DESKTOP_SESSION_ID: 'this-is-deprecated',
  GNOME_SHELL_SESSION_MODE: 'classic',
  GPG_AGENT_INFO: '/run/user/1000/gnupg/S.gpg-agent:0:1',
  GSETTINGS_SCHEMA_DIR: '/snap/code/13/usr/share/glib-2.0/schemas',
  GTK_MODULES: 'gail:atk-bridge',
  HOME: '/home/fhungenberg',
  LANG: 'de_DE.UTF-8',
  LOGNAME: 'fhungenberg',
  OLDPWD: '/home/fhungenberg/github/vscode-test',
  PATH: '/home/fhungenberg/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/snap/bin',
  PWD: '/home/fhungenberg/github/vscode-test',
  QT_ACCESSIBILITY: '1',
  SESSION_MANAGER: 'local/ws-fhungenberg:@/tmp/.ICE-unix/15662,unix/ws-fhungenberg:/tmp/.ICE-unix/15662',
  SHELL: '/usr/bin/zsh',
  SHLVL: '1',
  TERM_PROGRAM_VERSION: '1.37.1',
  TERM_PROGRAM: 'vscode',
  TERM: 'xterm-256color',
  USER: 'fhungenberg',
  USERNAME: 'fhungenberg',
  WINDOWPATH: '2',
  XAUTHORITY: '/run/user/1000/gdm/Xauthority',
  XDG_CACHE_HOME: '/home/fhungenberg/snap/code/common/.cache',
  XDG_CURRENT_DESKTOP: 'GNOME-Classic:GNOME',
  XDG_DATA_DIRS: '/usr/share/gnome:/usr/local/share:/usr/share:/var/lib/snapd/desktop',
  XDG_MENU_PREFIX: 'gnome-',
  XDG_RUNTIME_DIR: '/run/user/1000/snap.code',
  XDG_SEAT: 'seat0',
  XDG_SESSION_CLASS: 'user',
  XDG_SESSION_DESKTOP: 'gnome-classic',
  XDG_SESSION_ID: '2',
  XDG_SESSION_TYPE: 'x11',
  XDG_VTNR: '2',
}
```

From terminal with `node script.js`:

```
{
  _: '/home/fhungenberg/.nvm/versions/node/v12.9.1/bin/node',
  COLORTERM: 'truecolor',
  DBUS_SESSION_BUS_ADDRESS: 'unix:path=/run/user/1000/bus',
  DESKTOP_AUTOSTART_ID: '1043df04cec7af0dd156706103693144600000156620007',
  DESKTOP_SESSION: 'gnome-classic',
  DISPLAY: ':0',
  GDM_LANG: 'de_DE.UTF-8',
  GDMSESSION: 'gnome-classic',
  GIO_LAUNCHED_DESKTOP_FILE_PID: '18646',
  GIO_LAUNCHED_DESKTOP_FILE: '/etc/xdg/autostart/org.gnome.SettingsDaemon.MediaKeys.desktop',
  GNOME_DESKTOP_SESSION_ID: 'this-is-deprecated',
  GNOME_SHELL_SESSION_MODE: 'classic',
  GPG_AGENT_INFO: '/run/user/1000/gnupg/S.gpg-agent:0:1',
  GTK_MODULES: 'gail:atk-bridge',
  HOME: '/home/fhungenberg',
  LANG: 'de_DE.UTF-8',
  LESS: '-R',
  LOGNAME: 'fhungenberg',
  LS_COLORS: 'rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:',
  LSCOLORS: 'Gxfxcxdxbxegedabagacad',
  NVM_BIN: '/home/fhungenberg/.nvm/versions/node/v12.9.1/bin',
  NVM_CD_FLAGS: '-q',
  NVM_DIR: '/home/fhungenberg/.nvm',
  OLDPWD: '/home/fhungenberg/projects/docker-development',
  PAGER: 'less',
  PATH: '/snap/bin:/home/fhungenberg/.nvm/versions/node/v12.9.1/bin:/home/fhungenberg/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/snap/bin',
  PWD: '/home/fhungenberg/github/vscode-test',
  QT_ACCESSIBILITY: '1',
  SESSION_MANAGER: 'local/ws-fhungenberg:@/tmp/.ICE-unix/15662,unix/ws-fhungenberg:/tmp/.ICE-unix/15662',
  SHELL: '/usr/bin/zsh',
  SHLVL: '1',
  SPACESHIP_ROOT: '/home/fhungenberg/.oh-my-zsh/custom/themes/spaceship-prompt',
  SPACESHIP_VERSION: '3.11.1',
  TERM: 'xterm-256color',
  TERMINATOR_DBUS_NAME: 'net.tenshu.Terminator21a9d5db22c73a993ff0b42f64b396873',
  TERMINATOR_DBUS_PATH: '/net/tenshu/Terminator2',
  TERMINATOR_UUID: 'urn:uuid:e98b2d54-4b91-4a77-b732-563c7b8cd432',
  USER: 'fhungenberg',
  USERNAME: 'fhungenberg',
  VTE_VERSION: '5402',
  WINDOWPATH: '2',
  XAUTHORITY: '/run/user/1000/gdm/Xauthority',
  XDG_CURRENT_DESKTOP: 'GNOME-Classic:GNOME',
  XDG_DATA_DIRS: '/usr/share/gnome:/usr/local/share:/usr/share:/var/lib/snapd/desktop',
  XDG_MENU_PREFIX: 'gnome-',
  XDG_RUNTIME_DIR: '/run/user/1000',
  XDG_SEAT: 'seat0',
  XDG_SESSION_CLASS: 'user',
  XDG_SESSION_DESKTOP: 'gnome-classic',
  XDG_SESSION_ID: '2',
  XDG_SESSION_TYPE: 'x11',
  XDG_VTNR: '2',
  ZSH: '/home/fhungenberg/.oh-my-zsh',
}
```

**Comment by alexr00** [comment]
@shiftgeist how are you launching VS Code?

**Comment by shiftgeist** [comment]
Meta key - code
![image](https://user-images.githubusercontent.com/30648730/63956554-f1a40f00-ca86-11e9-895d-27045c34f2b4.png)

Installed via snap.

**Comment by alexr00** [comment]
I think that your environment isn't getting correctly set in VS Code when launching from the UI. It may be because it is installed via snap. 

I'm still trying to confirm that this is actually what is happening. In the mean time, could you uninstall and re-install not from snap and see if works?

**Comment by shiftgeist** [comment]
When running a npm script with in code, installed via a downloaded .deb file, everything works as expected. :+1: 

Seems this has something todo with the snap installation.
Maybe because of the `--classic` param at the code snap installation? (I have no understanding what that does.)

**Comment by alexr00** [comment]
@joaomoreno it's something to do with the snap.

@shiftgeist in the snap installation can you open the dev tools then try running `getLazyEnv().then(e => console.log(e))`

**Comment by joaomoreno** [comment]
Snap + env = 🐛  https://github.com/microsoft/vscode/issues/72042

**Comment by shiftgeist** [comment]
Output of `getLazyEnv()...`

```
APPLICATION_INSIGHTS_NO_DIAGNOSTIC_CHANNEL: "true"
COLORTERM: "truecolor"
DBUS_SESSION_BUS_ADDRESS: "unix:path=/run/user/1000/bus"
DESKTOP_AUTOSTART_ID: "10eb4188cbfc5651cb156715829813855700000145250007"
DESKTOP_SESSION: "gnome-classic"
DISABLE_WAYLAND: "1"
DISPLAY: ":0"
ELECTRON_NO_ATTACH_CONSOLE: "1"
GDK_PIXBUF_MODULEDIR: "/snap/code/13/usr/lib/x86_64-linux-gnu/gdk-pixbuf-2.0/2.10.0/loaders"
GDK_PIXBUF_MODULE_FILE: "/home/fhungenberg/snap/code/common/.cache/gdk-pixbuf-loaders.cache"
GDMSESSION: "gnome-classic"
GDM_LANG: "de_DE.UTF-8"
GIO_LAUNCHED_DESKTOP_FILE: "/etc/xdg/autostart/org.gnome.SettingsDaemon.MediaKeys.desktop"
GIO_LAUNCHED_DESKTOP_FILE_PID: "16453"
GNOME_DESKTOP_SESSION_ID: "this-is-deprecated"
GNOME_SHELL_SESSION_MODE: "classic"
GPG_AGENT_INFO: "/run/user/1000/gnupg/S.gpg-agent:0:1"
GSETTINGS_SCHEMA_DIR: "/snap/code/13/usr/share/glib-2.0/schemas"
GTK_MODULES: "gail:atk-bridge"
HOME: "/home/fhungenberg"
LANG: "de_DE.UTF-8"
LESS: "-R"
LOGNAME: "fhungenberg"
LSCOLORS: "Gxfxcxdxbxegedabagacad"
LS_COLORS: "rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:"
NVM_BIN: "/home/fhungenberg/.nvm/versions/node/v12.9.1/bin"
NVM_CD_FLAGS: "-q"
NVM_DIR: "/home/fhungenberg/.nvm"
OLDPWD: "/home/fhungenberg/projects/docker-development"
PAGER: "less"
PATH: "/snap/bin:/snap/bin:/snap/bin:/home/fhungenberg/.nvm/versions/node/v12.9.1/bin:/home/fhungenberg/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/snap/bin"
PWD: "/home/fhungenberg/projects/docker-development"
QT_ACCESSIBILITY: "1"
SESSION_MANAGER: "local/ws-fhungenberg:@/tmp/.ICE-unix/14525,unix/ws-fhungenberg:/tmp/.ICE-unix/14525"
SHELL: "/usr/bin/zsh"
SHLVL: "4"
SNAP: "/snap/code/13"
SNAP_ARCH: "amd64"
SNAP_COMMON: "/var/snap/code/common"
SNAP_CONTEXT: "3Ltgc7V85ltH8sySNJnFY2fs1DTK31rLxrtH5wsJTpFL"
SNAP_COOKIE: "3Ltgc7V85ltH8sySNJnFY2fs1DTK31rLxrtH5wsJTpFL"
SNAP_DATA: "/var/snap/code/13"
SNAP_INSTANCE_KEY: ""
SNAP_INSTANCE_NAME: "code"
SNAP_LIBRARY_PATH: "/var/lib/snapd/lib/gl:/var/lib/snapd/lib/gl32:/var/lib/snapd/void"
SNAP_NAME: "code"
SNAP_REEXEC: ""
SNAP_REVISION: "13"
SNAP_USER_COMMON: "/home/fhungenberg/snap/code/common"
SNAP_USER_DATA: "/home/fhungenberg/snap/code/13"
SNAP_VERSION: "f06011ac"
SPACESHIP_ROOT: "/home/fhungenberg/.oh-my-zsh/custom/themes/spaceship-prompt"
SPACESHIP_VERSION: "3.11.1"
SSH_AGENT_PID: "14524"
SSH_AUTH_SOCK: "/run/user/1000/keyring/ssh"
TERM: "xterm-256color"
TERMINATOR_DBUS_NAME: "net.tenshu.Terminator21a9d5db22c73a993ff0b42f64b396873"
TERMINATOR_DBUS_PATH: "/net/tenshu/Terminator2"
TERMINATOR_UUID: "urn:uuid:f1322d84-712a-45d9-9d73-3e96743d26b6"
USER: "fhungenberg"
USERNAME: "fhungenberg"
VSCODE_CLI: "1"
VSCODE_IPC_HOOK: "/run/user/1000/snap.code/vscode-94e24a67-1.37.1-main.sock"
VSCODE_LOGS: "/home/fhungenberg/.config/Code/logs/20190830T142613"
VSCODE_NLS_CONFIG: "{"locale":"de","availableLanguages":{},"_languagePackSupport":true}"
VTE_VERSION: "5402"
WINDOWPATH: "2"
XAUTHORITY: "/run/user/1000/gdm/Xauthority"
XDG_CACHE_HOME: "/home/fhungenberg/snap/code/common/.cache"
XDG_CURRENT_DESKTOP: "GNOME-Classic:GNOME"
XDG_DATA_DIRS: "/usr/share/gnome:/usr/local/share:/usr/share:/var/lib/snapd/desktop"
XDG_MENU_PREFIX: "gnome-"
XDG_RUNTIME_DIR: "/run/user/1000/snap.code"
XDG_SEAT: "seat0"
XDG_SESSION_CLASS: "user"
XDG_SESSION_DESKTOP: "gnome-classic"
XDG_SESSION_ID: "2"
XDG_SESSION_TYPE: "x11"
XDG_VTNR: "2"
ZSH: "/home/fhungenberg/.oh-my-zsh"
_: "/snap/code/13/usr/share/code/bin/../code"
```

**Comment by alexr00** [comment]
Should be fixed as a result of #72042!
